Contributor's Guide draft (#1824)
FluxML/Flux.jl: CONTRIBUTING.md

# Flux Contributor’s Guide

Welcome to the Flux community! 

**Flux.jl** is the de facto machine learning library in the **Julia ecosystem** and one of the foundational Julia packages. It is written in *pure Julia*. This is very different from most other state-of-the-art machine learning libraries which typically leverage languages like C or C++ under the hood.

Flux has always had the mission of being a simple, hackable and performant approach to machine learning, which is extended to a number of domains in science by means of differentiable programming. 

> Julia provides features that Machine Learning researchers and practitioners need. However, the ecosystem is not yet mature enough so that developers can rewrite it in Julia. This provides a great opportunity for you to contribute to the [future of Machine Learning and the Julia language in general](https://towardsdatascience.com/the-future-of-machine-learning-and-why-it-looks-a-lot-like-julia-a0e26b51f6a6). 

Reviewer: I would reword it to say the ecosystem is rapidly growing, but maybe we can remove the links to the blog posts. Is it common for contribution guides to have such links in them?

In this guide, you can find information on how to make your first contribution to Flux, some ideas for getting started, and some guidelines on how to submit a contribution. 

## Why contribute to Flux?

You can make a difference to one of the most quickly growing Deep Learning frameworks, the Julia Language, and the future of open-source software as a whole. Open source projects rely on contributions from volunteers. Contributions enable both the project and volunteers to grow and develop. No matter how you contribute to Flux, it will be a great experience for you.

Contributing to Flux brings several benefits to you and the community such as:

> * Contribute to the cause of open and reproducible science.
> * Help grow the adoption of the FluxML ecosystem.
> * Become a member of a community that’s excited about open source and sharing knowledge.
> * Polish your teaching skills by helping others get started using Flux.
> * Showcase your work by creating examples that others can use in their projects.
> * Build a track record of public contributions which will help build your career.

Reviewer 2: ```suggestion
* Contribute to the cause of open and reproducible science
* Help grow the adoption of the FluxML ecosystem
* Become a member of a community that’s excited about open source and sharing knowledge
* Polish your teaching skills by helping others get started using Flux
* Showcase your work by creating examples that others can use in their projects
* Build a track record of public contributions which will help build your career
```

## Tools

Before you start contributing to Flux, you need the following tools:

* **Julia:** For more information on how to install Julia, see [Download Julia (in under 2.5 minutes)](https://www.youtube.com/watch?v=t67TGcf4SmM).
* **IDE:** You can set one of the following IDEs as well as their extensions for developing in Julia:

>    * [Julia for Visual Studio Code](https://www.julia-vscode.org/)

Reviewer: I don't think we need to say this here.

   * [IJulia](https://github.com/JuliaLang/IJulia.jl)
   * [Pluto.jl](https://github.com/fonsp/Pluto.jl)
   * [Julia-vim](https://github.com/JuliaEditorSupport/julia-vim)
   * [Emacs major mode for the Julia programming language](https://github.com/JuliaEditorSupport/julia-emacs)
   * [Julia language support for Notepad++](https://github.com/JuliaEditorSupport/julia-NotepadPlusPlus)
* **Knowledge of Git and how to create a pull request:** For more information on getting started with Git, see [Making a first Julia pull request](https://kshyatt.github.io/post/firstjuliapr/).

## Learn Flux

If you need to learn about Julia and Flux, then you can check out the following resources:

> * [JuliaAcademy](https://juliaacademy.com/) introductory courses to Julia and Flux.
> * Flux’s official [Getting Started](https://fluxml.ai/getting_started.html) tutorial.
> * [Deep Learning with Flux - A 60 Minute Blitz](https://fluxml.ai/tutorials/2020/09/15/deep-learning-flux.html). This is a quick intro to Flux loosely based on PyTorch’s tutorial.
> * [Flux Model Zoo](https://github.com/FluxML/model-zoo) showcases various demonstrations of models that you can reuse with your own data.
> * [Flux’s official documentation](https://fluxml.ai/Flux.jl/stable/).

Reviewer 2: ```suggestion
* [JuliaAcademy](https://juliaacademy.com/) introductory courses to Julia and Flux
* Flux’s official [Getting Started](https://fluxml.ai/getting_started.html) tutorial
* [Deep Learning with Flux - A 60 Minute Blitz](https://fluxml.ai/tutorials/2020/09/15/deep-learning-flux.html): a quick intro to Flux loosely based on PyTorch’s tutorial
* [Flux Model Zoo](https://github.com/FluxML/model-zoo) showcases various demonstrations of models that you can reuse with your own data
* [Flux’s official documentation](https://fluxml.ai/Flux.jl/stable/)
```

## Get help

The Flux community is more than happy to help you with any questions that you might have related to your contribution. You can get in touch with the community in any of the following channels:

> * **Stack Overflow:** Search for [Flux/ ML](https://stackoverflow.com/questions/tagged/flux.jl) tags.
> * **Discourse forum:** Go to [Machine Learning in Julia community](https://discourse.julialang.org/c/domain/ml/24).
> * **Events:** Attend the [ML bi-weekly calls](https://julialang.org/community/#events).
> * **Community team:** Join the group of [community maintainers](https://github.com/FluxML/ML-Coordination-Tracker) supporting the FluxML ecosystem.
> * **Slack:** Join the [Official Julia Slack](https://julialang.org/slack/) for casual conversation.
> * **Zulip:** Join the [Zulip Server for the Julia programming language](https://julialang.zulipchat.com/login/) community.

Reviewer 2: Just reordering based on importance and added link to Zulip stream

```suggestion
* **Discourse forum:** Go to [Machine Learning in Julia community](https://discourse.julialang.org/c/domain/ml/24)
* **Community team:** Join the group of [community maintainers](https://github.com/FluxML/ML-Coordination-Tracker) supporting the FluxML ecosystem (see the [Zulip stream](https://julialang.zulipchat.com/#narrow/stream/237432-ml-ecosystem-coordination) as well)
* **Slack:** Join the [Official Julia Slack](https://julialang.org/slack/) for casual conversation (see `#flux-bridged` and `#machine-learning`)
* **Zulip:** Join the [Zulip Server for the Julia programming language](https://julialang.zulipchat.com/login/) community
* **Stack Overflow:** Search for [Flux/ ML](https://stackoverflow.com/questions/tagged/flux.jl) tags
* **Events:** Attend the [ML bi-weekly calls](https://julialang.org/community/#events)
```

Reviewer: Pretty sure the "Get Help" section should start with places to get existing help from. In these cases Discourse and SO would be more important followed by Slack, Zulip and then the ML Tracker.

Reviewer 2: The reason I put SO lower is cause no devs seem to consistently check it. So I ordered based on how quickly they might get help from fellow contributors.

No matter if other people have already written about the same topic before. You can always bring a new point of view, dig deeper into a topic, or provide a new use case using a different data set. For more information on how to create a pull request with your tutorial, see [Flux website README](https://github.com/FluxML/fluxml.github.io).

### Other

We are open to contributions in all shapes and forms! If you have an idea, please [suggest it](#get-help) to the community and we will do our best to help you bring it to life!

## Contribution guidelines

Flux.jl and its ecosystem follow the ColPrac: Contributor's Guide on Collaborative Practises for Community Packages guide. For more information about best practises on contributing to packages, see the [ColPrac guide](https://github.com/SciML/ColPrac).

We also suggest taking a look at the [Julia Ecosystem Contributor’s Guide](https://julialang.org/contribute/) guide which goes over higher-level topics like why you should contribute to the Julia ecosystem, different contribution pathways, and more!

In this section, you can find information on the following:

* [How to file a bug report]()
* [How to submit a contribution]()

> * [Flux source code organisation]()

Reviewer: Should link to the appropriate sections.

Author: Links fixed. Just a question: does it makes sense to have the section Flux source code organisation? I haven't finished that section and just wondering if it's worth it.

* **Events:** Attend the [ML bi-weekly calls](https://julialang.org/community/#events)

## How to contribute to Flux
In this section, you can find a step by step guide to make your first Flux contribution. These are the suggested steps for a complete beginner but feel free to adapt them to your own style and journey.

Before you begin working on your contribution, make sure you have installed Julia and an IDE (see section [Tools](#tools)). In case of any issues, remember you can always [get help](#get-help) from the community. 

To make your first contribution to Flux:

1. Read about getting started with **Git** and **how to contribute to an open-source project**. The tutorial [Making a first Julia pull request](https://kshyatt.github.io/post/firstjuliapr/) is a great starting point for learning about git and how to contribute to the Julia language in general. 
2. Read the [Contribution ideas](#contribution-ideas) section to get inspiration.
3. Set up your local environment. Depending on your contribution, fork and clone the Flux, Model Zoo or Flux website GitHub repositories. 
4. Build Flux locally.

> 5. Familiarise yourself with the Flux source code. For more information on the source code organisation, see section….

Reviewer 2: ```suggestion
5. Familiarise yourself with the Flux source code. See [Flux source code organization](#flux-source-code-organization) for more information.
```

* The version of Julia as provided by the versioninfo() command. Occasionally, the longer output produced by versioninfo(verbose = true) may be useful also, especially if the issue is related to a specific package.

> **Note:** When pasting code blocks or output, put triple back quotes (```) around the text so GitHub will format it nicely. Code statements should be surrounded by single backquotes (`). Be aware that the @ sign tags users on GitHub, so references to macros should always be in single back quotes. See [GitHub's guide on Markdown](https://guides.github.com/features/mastering-markdown) for more formatting tricks.

### How to submit a contribution

You can use [this template](https://github.com/FluxML/Flux.jl/blob/master/.github/pull_request_template.md) as a starting point when creating a new pull request. 

### Flux source code organization

The following table shows how the Flux code is organized:

| Directory  | Contents |
| ------------- | ------------- |
| docs  | Documentation site  |
| paper  | Paper that describes Flux |

> | perf  |    |

Reviewer 2: For now I would leave this out, since it is kind of a dead branch.

Author: Thank you so much @darsnack for all of your comments!

Reviewer 2: No problem! Just to clarify, I meant delete the row for `perf` only. I think the section you added on source code organization is a great idea.

Author: Ah! Sorry for the confusion. I added the table :)
